Add tests for GitHub issues on the "Zapoj se" page

honzajavorek opened the pull request:
- 83b0d14 makes the application a Python package (see commit message on why)
- 4350887
  - renames several identifiers in `github.py`
  - fixes some minor errors, adds `github_test.py` with tests
  - adds `github_test_fixtures.py` with helpers for generating some random fixtures resembling possible data coming from the GitHub API v4
- da18544 adds tests coverage

This PR _does not_ test or solve https://github.com/pyvec/python.cz/issues/191, that would be a next step. I haven't been writing so much code and so much tests in Python for a long time, so any kind of feedback would be greatly appreciated.

-----

⚠️ This changes contents of `deployment/update.sh`, so the deploy needs to be done manually when the PR gets merged.

hroncok commented on requirements.txt:
```diff
@@ -1,8 +0,0 @@
-czech-sort==0.4
-PyYAML==3.12
-python-slugify==1.2.1
-responses==0.7.0
```
This file is still referenced in the documentation (README). Please update the docs according to change in this commit (and preferably squash the change to this commit).

hroncok replied:
@honzajavorek Note that this is shown as outdated, but is relevant.

hroncok commented on README.md:
```diff
@@ -47,3 +47,9 @@ The site is backed by [Pyvec](http://pyvec.org/), nonprofit organization dedicat
+.. |Build Status| image:: https://travis-ci.org/pyvec/python.cz.svg?branch=master
+   :target: https://travis-ci.org/pyvec/python.cz
+.. |Test Coverage| image:: https://coveralls.io/repos/github/pyvec/python.cz/badge.svg?branch=master
+   :target: https://coveralls.io/github/pyvec/python.cz?branch=master
```
This is RST in Markdown file. It doesn't work.

honzajavorek replied:
😱

hroncok commented on .travis.yml:
```diff
@@ -3,10 +3,12 @@ python:
 script:
     - "flake8 ."
-    - "py.test pythoncz_tests"
+    - "pytest --cov-report="
```
I see no reasonable output in the Travis CI build logs. Is it desired? I know that coveralls will update the PRs etc., but I would like to be able to see the information when i looks at the Travis CI log.

honzajavorek replied:
I can remove the `--cov-report=`, hopefully that wouldn't affect coveralls. I have it there in [fiobank](https://github.com/honzajavorek/fiobank) for some reason, but I don't remember why.

hroncok commented on .travis.yml:
```diff
@@ -3,10 +3,12 @@ python:
-    - "pip install ."
+    - "pip install .[tests]"
 script:
     - "flake8 ."
```
Not in the scope of this PR, but maybe later: Consider using pytest-flake8 here.

hroncok commented:
(Note that I've accidentally skipped the main commit (oops) when reviewing, so more feedback might be on it's way.)

hroncok commented on pythoncz_tests/models/github_test.py:
```diff
@@ -0,0 +1,432 @@
+        for issue in repo['issues']['nodes']:
+            expected_titles_are_pr[issue['title']] = False
+        for pull_request in repo['pullRequests']['nodes']:
+            expected_titles_are_pr[pull_request['title']] = True
```
A dict comprehension might be nicer?

Somehting like:

```python
repos = {**repo1, **repo2}
issues = {issue['title']: False for issue in repos['issues']['nodes']}
pull_requests = {pr['title']: True for pr in repos['pullRequests']['nodes']}
expected_titles_are_pr = {**issues, **pull_requests}
```

Ok, might not be nicer at all. I'm not sure. Whatever you like :D

honzajavorek replied:
Man, you must be coming from future! What is this?! 🙈 

![back-to-the-future-honest-trailer](https://user-images.githubusercontent.com/283441/29324117-517844e4-81e3-11e7-8c10-09024b85b9d4.jpg)

honzajavorek replied:
I decided to leave this as it is. It's verbose, but quite clear.

honzajavorek replied:
I don't mind it's new, it just felt to me this particular case isn't much more readable with the comprehensions.

But ❗️ the production version is 3.4 so this won't work on production 😱 @hroncok And we accepted the other comprehension.

hroncok replied:
>  the production version is 3.4

And so is Travis and my venv. So we don't need to worry about pushing backwards incompatible code.

hroncok replied:
The incompatible code here is not dict comprehension, but `{**a, **b}`

honzajavorek replied:
Ah! All right then.

messa replied:
We have sent our comments just seconds apart, I didn't know we will keep the first variant. 

Yes, only the code `{**a, **b}` is incompatible: https://docs.python.org/3/whatsnew/3.5.html#pep-448-additional-unpacking-generalizations

hroncok commented on pythoncz_tests/models/github_test.py:
```diff
@@ -0,0 +1,432 @@
+
+    titles_are_pr = {}
+    for issue in issues:
+        titles_are_pr[issue['title']] = issue['is_pull_request']
```
But this one would definitively be better as comprehension:

```python
titles_are_pr = {i['title']: i['is_pull_request'] for i in issues}
```

hroncok commented on pythoncz_tests/models/github_test.py:
```diff
@@ -0,0 +1,432 @@
+    assert excinfo.value.response.status_code == 200
+
+
+def test_request_api_500_errors(github, requests_mock):
```
Parametrize with the previous one.

honzajavorek replied:
Does parametrize handle assert vs. raises?

hroncok replied:
Both the tests are raises + assert.

hroncok commented on pythoncz_tests/models/github_test.py:
```diff
@@ -0,0 +1,432 @@
+    issues = [github._format_issue('org', repository, issue) for issue in [
+        fixtures.issue(labels=[], reactions_counts={'LAUGH': 1}),
+        fixtures.issue(labels=[], reactions_counts={'THUMBS_DOWN': 1}),
+        fixtures.issue(labels=[], reactions_counts={'HEART': 5, 'CONFUSED': 1}),  # NOQA
```
You are testing that `{'HEART': 5, 'CONFUSED': 1}` is 4 in another test. Keep this one simple with `{'THUMBS_UP': 4}` - you'll get rid of the long line and it will be a bit more clear that if this test fails, it's probably the sorting and not the counting.

hroncok commented on pythoncz_tests/models/github_test.py:
```diff
@@ -0,0 +1,432 @@
+
+def test_sort_issues_activity(github):
+    """
+    Issues with most user activity should go first if there's no 'coach'
```
It's not clear from the docstring that this is sorted by the sum of comment count and participants count. Also make sure that comments only would sort differently (because here `3 > 1 > 0`), so the test would pass if the only criterion would be comments count.

hroncok replied:
Fixed.

messa commented on pythoncz/models/github.py:
```diff
@@ -125,23 +136,26 @@ def _get_nodes(node, connection_name):
 
-        if connection_nodes_count > connection_nodes_total_count:
+        if connection_nodes_count < connection_nodes_total_count:
             warnings.warn((
```
This is "runtime warning", not warning about usage of obsoleted fuction or similar. This should be handled by logging, not warnings.

honzajavorek replied:
The docs say

> Warning messages are typically issued in situations where it is useful to alert the user of some condition in a program, where that condition (normally) doesn’t warrant raising an exception and terminating the program.

I agree in web apps this is usually typically handled by logging, but there's no logging in `pythoncz` yet. Also, nobody reads the production logs 😄 So in most cases, the warning would appear in a terminal to someone who develops the site locally. When spotted, the developer should file an issue on this repo and the problem should be solved by raising limits or pagination.

messa commented on pythoncz_tests/models/github_test.py:
```diff
@@ -0,0 +1,432 @@
+from werkzeug.contrib.cache import NullCache
+
+from pythoncz.models import github as github_module
+from pythoncz_tests.models import github_test_fixtures as fixtures
```
Why not use pytest fixtures? I don't think this is bad practice but I've already been there and then refactored it as pytest.fixtures and the code got much cleaner :)

honzajavorek replied:
I didn't find out how to pass arguments to pytest fixtures. I had them, then realized I'm probably overcomplicating things and then I decided to make them just simple functions, taking arguments and producing fixtures. If you know about a way how to make the tests cleaner and if turning the module into pytest fixtures contributes to that goal, go for it 👍 But I suggest that as a next step to this PR.

hroncok replied:
Those are pretty much functions. Sometimes called multiple times in a test. Not sure if pytest fixtures would be a good fit.

messa commented on pythoncz_tests/models/github_test.py:
```diff
@@ -0,0 +1,432 @@
+    """Provides mechanism for faking HTTP responses"""
+    with responses.RequestsMock() as mock:
+        yield mock
+    mock.reset()
```
Why do you call `mock.reset()` here? It doesn't make sense to having to call some reset after with block - and indeed it is already included in `RequestsMock.__exit__()`: https://github.com/getsentry/responses/blob/master/responses.py#L381

honzajavorek replied:
It did not work for me without it, but when I recall the problems I had, I think it might have been because of a different issue. If now the tests work even without this line, then it's just a leftover and let's remove it.

honzajavorek replied:
Just tried it and it works without the reset. I'll remove it.

messa commented on pythoncz_tests/models/github_test.py:
```diff
@@ -0,0 +1,432 @@
+    """Provides the 'github' module with caching disabled"""
+    original_cache = github_module.cache
+    github_module.cache = NullCache()
+    yield github_module
```
Btw. pytest fixtures call the code after `yield` even if the test itself fails. But of course using proper resource management (with-block or try-finally) is always a good practice.

hroncok replied:
@messa didn't know that.

hroncok replied:
Also on errored tests?

honzajavorek replied:
Yup, this is special pytest behavior. You write tear down code after `yield` and that's it. Not sure about E tests, would need to see docs.

hroncok replied:
Just have checked. Even on E tests. TIL.

honzajavorek commented on pythoncz/models/github.py:
```diff
@@ -81,6 +71,27 @@ def _get_issues_for_org(session, org_name):
+        raise ValueError(
+            'Unexpected structure of the GitHub API response: {}'.format(e)
+        )
+    else:
```
Reading of the code was more straightforward this way for me, and also I've been changing these lines many times (alongside with testing them...), so it might as well be just a leftover. I have no strong opinions about it, I'm completely okay to remove the `else`.

hroncok replied:
No strong opinions here either. Slightly inclined to remove it, as I would change it in this case:

```python
def foo():
    if whatnot():
        ...
        return 123
    else:
        ...
        return 563
```

To:


```python
def foo():
    if whatnot():
        ...
        return 123
    ...
    return 563
```

honzajavorek commented on pythoncz_tests/models/github_test.py:
```diff
@@ -0,0 +1,432 @@
+    issue_org1 = fixtures.issue()
+    requests_mock.add(
+        responses.POST,
+        'https://api.github.com/graphql',
```
I have bad experiences with DRYing tests too much, because it heavily affects readability and clarity of the test case. It's DRYier, but it requires many context switches for the reader to get the whole point of the test. So in tests I'm usually inclining to rather repeat myself. But this might be worth DRYing if the naming is done right, I agree. I'll think about it.

honzajavorek commented:
@hroncok @messa Thanks for the reviews! Amazing work! And I'm deeply sorry for issuing a PR so large. I'll get better in making smaller bits, I promise. It might take me a moment to get back to this, but I'll address your comments and let you know 🚀

honzajavorek commented:
(It's actually so awesome anyone is doing these reviews that I'm almost getting emotional. I remember when I was coding first versions of this site and after waiting for a review for several weeks I was told to just develop the thing and deploy it, not to bother people with reviews 😄)

hroncok commented:
@honzajavorek if you put thumbps up to everything you agree unconditionally, I'll send a PR to you with changes ;)

honzajavorek commented:
- 👍 in reactions = I agree
- agreeing or being indifferent in comments = I agree
- otherwise I'd like to keep thinking about the changes (DRYing, also the first dict comprehensions, which look a bit intimidating)

honzajavorek commented:
Note for myself/ourselves - do not forget about [this](https://github.com/pyvec/python.cz/pull/211#discussion_r133158939), GitHub hides it by default.

hroncok commented:
@honzajavorek #212

honzajavorek commented:
@hroncok You've rebased this? I've added something to previous commits as well locally... I guess I'm screwed now.

hroncok commented:
No, I haven't. I merely merged with a merge commit.

honzajavorek commented:
Hm, then I don't know. But I can't update my local copy now.

hroncok commented:
Show me your git log.

honzajavorek commented:
I've added some changes, commited them, then rebased my branch and added this new commit to the "Added tests for GitHub issues" commit. Since then I'm getting conflicts when I try `git pull --rebase origin honzajavorek/zapojse-tests`.

I've already destroyed the branch so I could create [honzajavorek/zapojse-tests-screwed-commit](https://github.com/pyvec/python.cz/tree/honzajavorek/zapojse-tests-screwed-commit) with just the modified commit, so there's no more `git log` for the original branch.

honzajavorek commented:
Basically I could take this commit and cherry-pick all the subsequent ones and to force push everything here 🤔 I'm going to do it.

honzajavorek commented:
Everything should be sorted now and all addressed, except of the fact the code should blow up on Python 3.4 if @messa is right about which version supports the dict comprehensions. OTOH, it should then blow up on Travis CI as well... 🤔

honzajavorek commented:
Okay, I think we're good to go per this comment https://github.com/pyvec/python.cz/pull/211#discussion_r133288347. @hroncok will you merge? I can deploy afterwards (manual deploy is needed for this one).

hroncok commented:
> Next time we should be more cautious about squashing and rebasing and squashing and rebasing and working in multiple people over the same branch.

 1. only add fixup commits during the entire thing, never rebase
 2. once approved, autosquash, push force and merge

honzajavorek commented:
Good point. When consolidating the branch, I realized you made change on a similar line as me. Even though for a different reason and by human eye it was clearly unrelated to each other, git thought it's a conflict. The steps you describe would help us to avoid it.

honzajavorek commented:
Deployed 🚀

hroncok commented:
Thanks for the tests.
